Add status.js and its tests

AnkshitJain opened the pull request:
@prasadtalasila Please review the commit and suggest the necessary changes. I have included a `environmentCheck.js` file in `/util` directory that checks whether an environment variable is defined or not.
This module can be used in other js modules as well.

prasadtalasila commented on load_balancer/status.js:
```diff
@@ -0,0 +1,105 @@
+const logger = new Logger(loggerConfig);
+
+function getComponentStatus(node) {
+  return new Promise((resolve) => {
```
A coding rule known as step down rule from Robert Martin's Clean Code book will certainly help get a good structure. In step down rule, if f2() calls f1(), then code for f2() comes first in the file followed by f1().

AnkshitJain commented on util/environmentCheck.js:
```diff
@@ -0,0 +1,9 @@
+};
+
+module.exports = {
+  check,
```
The [airbnb](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/style.js#L41) config file that we extend for eslint has this convention. The [rule](https://eslint.org/docs/rules/comma-dangle) at eslint says we should be consistent in what we intend to follow in this case.

prasadtalasila replied:
sounds good.

AnkshitJain commented:
@prasadtalasila Please have a look at the updated commit.

prasadtalasila commented on load_balancer/load_balancer.js:
```diff
@@ -320,57 +286,20 @@ try {
 {
-  server.listen(nodes_data.host_port.port);
-  console.log("Listening at "+nodes_data.host_port.port);
+  server.listen(nodes_data.load_balancer.port);
```
please fix the code quality issues raised here.

AnkshitJain replied:
The check doesn't show any new errors here. Can you tell what the error is?

prasadtalasila replied:
You may need to install the codeclimate plugin on chrome / chromium browser to have the codeclimate issues show up on the PR page.

AnkshitJain replied:
I expect 4 errors here:
1. Opening curly brace does not appear on the same line as controlling statement.
2. Strings must use single quote.
3. Unexpected string concatenation.
4. Infix operators must be spaced.
These are previously existing quality errors, and I haven't tried to fix them. The only reason this line has a change is because the host name property was added there and the object in the configuration file nodes_data_config.json had the property name changed.

prasadtalasila replied:
A few more issues exist. I am not able to copy paste the consolidated list.

prasadtalasila commented on load_balancer/load_balancer.js:
```diff
@@ -30,56 +33,19 @@ app.get('/userCheck', function (req,res) {
+  lbStatus.status = 'up';
+  status.checkStatus((result) => {
+    //since the request is being processed, it is assumed that load balancer is up
+    const statusJson = result;
```
This looks fine, but there is a better library named [lodash](https://lodash.com/) that is good at manipulating json objects. You might want to use it.

prasadtalasila commented on load_balancer/status.js:
```diff
@@ -0,0 +1,102 @@
+/* eslint import/no-dynamic-require: 0 */
```
looks like this module can make use of lodash module to simplify the logic. Please see if you can use it to reduce the code.

prasadtalasila replied:
May be its just me, but the code does not look readable to me. Is there a way to simplify the code?

prasadtalasila replied:
I reviewed the lodash vs es6 debate. Its definitely worth using lodash in the project. I would rather introduce lodash more strategically by identifying the features of lodash that help us, note them down for use during refactoring. Nothing much to do now.

AnkshitJain commented on deploy/configs/load_balancer/nodes_data_conf.json:
```diff
@@ -1,22 +1,27 @@
 {
```
Do you suggest making changes to all the execution node configuration files i.e. conf.json and scores.json?

prasadtalasila replied:
Also to the configuration files of main_server as well.

AnkshitJain commented:
@prasadtalasila Please have a look at the updated commit.

prasadtalasila commented on load_balancer/status.js:
```diff
@@ -0,0 +1,90 @@
+    this.nodes = nodes;
+    this.promises = [];
+    for (let i = 0; i < this.nodes.length; i += 1) {
+      this.promises.push(getComponentStatus(this.nodes[i]));
```
At best, the code can resolve status of each component only once.

prasadtalasila commented on load_balancer/status.js:
```diff
@@ -0,0 +1,90 @@
+      resolve('down');
+    } else if (body.toString() === 'true') {
+      logger.debug(`Node at ${node.hostname}:${node.port} is up and running.`);
+      resolve('up');
```
Is the name of the callback function consistent with the nomeclature of promises?

AnkshitJain replied:
Yes, the promises callback are resolve and reject.

AnkshitJain commented on deploy/playbook.yml:
```diff
@@ -150,6 +150,7 @@
           - "../log/load_balancer:/log"
         env:
           LOGGERCONFIG: "/etc/util/logger.json"
+          LBCONFIG: "/etc/load_balancer/nodes_data_conf.json"
```
The `LOGGERCONFIG` variable would be used by all the components and is included as an environment variable for main_server, load_balancer and execution_nodes containers. The same cannot be said for the `LBCONFIG` variable. It has the node details and that is not required by the other containers.

prasadtalasila replied:
I did not ask for adding LBCONFIG environment variable to other containers. I suggested adding mainserver configuration variable to mainserver, execution node configuration variable to execution node. Done that way, there would not be any code replacements in the tests.

AnkshitJain commented:
@prasadtalasila The suggested structure of having the `Status` class before the module private methods gives an [error](https://eslint.org/docs/rules/no-use-before-define) according to our eslint conventions. Should this error be ignored or we stick with having method declarations at the top?

prasadtalasila commented:
@AnkshitJain Then lets follow the established conventions of ES6. Let's ignore the step down rule for now.

prasadtalasila commented:
@AnkshitJain async/await looks good. Please use the same.

AnkshitJain commented:
@prasadtalasila Please have a look at the updated commit. Code climate gives an error related to the code complexity. Initially when the method getComponentStatus was in two separate parts, it gave 2 errors saying the code was similar. Can you suggest what can be done in this case?

prasadtalasila commented:
@AnkshitJain Please modify the README.md file by updating the codeclimate badge to the following code.
```
[![Maintainability](https://api.codeclimate.com/v1/badges/e1668db2aeb3196311bc/maintainability)](https://codeclimate.com/github/AutolabJS/AutolabJS/maintainability)
```
Right next to codeclimate code quality badge, you can put the code coverage badge.
```
[![Test Coverage](https://api.codeclimate.com/v1/badges/e1668db2aeb3196311bc/test_coverage)](https://codeclimate.com/github/AutolabJS/AutolabJS/test_coverage)
```
Right now, this badge shows a question mark until we put in a code coverage tool. That's okay.

prasadtalasila commented on load_balancer/status.js:
```diff
@@ -0,0 +1,75 @@
+
+const logger = new Logger(loggerConfig);
+
+async function getComponentStatus(node) {
```
How about the following code structure? (BTW, there is a compilation error in the code snippet, please only use the idea, not the actual code)

```javascript
function isValidNode(node) {
        hostname = node.hostname;
        port = node.port;
        role = node.role;

        if (hostname === undefined || port === undefined || role == undefined)
                return false;
        else
                return true;
}

async function getComponentStatus(node) {
        if (!isValidNode(node))
                throw new InvalidNodeException();

        const resultJson = Object.assign({}, node);
        try {
                const body = await request.get(`https://${node.hostname}:${node.port}/connectionCheck`);
                if (body.toString() === 'true') {
                        logger.debug(`Node at ${node.hostname}:${node.port} is up and running.`);
                resultJson.status = 'up';
            }
        } catch (err) {
                logger.error(`Error connecting to ${node.hostname}:${node.port}`);
                logger.error(err);
                resultJson.status = 'down';
}
```

You also need to define InvalidNodeException class. I see the value in putting both isNodeValid() and InvalidNodeException in a module inside util/. That way, other components can cross check their configuration values as well.

AnkshitJain replied:
What properties wil the error class would require apart from an error message?

prasadtalasila replied:
just the error status message is sufficient.

prasadtalasila commented on load_balancer/status.js:
```diff
@@ -0,0 +1,75 @@
+  return resultJson;
+}
+
+async function getPromises(nodes) {
```
The previous name of this function sounds better. getPromises() name is too generic and unintuitive.

prasadtalasila commented on main_server/main_server.js:
```diff
@@ -28,10 +29,35 @@ var session = require('express-session')({
+    Path in docker containers: "/etc/main_server/APIKeys.json"
+*/
+
+check('MSCONFIG');
```
The check function is checking the definition of the environment variable. It can also check for the existence of the file and the read permissions. Once the existence and permissions checks pass, we can subject the configuration file to validity check by iterating the isNodeValid() function over all the entries in the configuration file.

This is just an idea. Please suggest alternatives and improvements.

prasadtalasila replied:
@AnkshitJain 
The scope of the above suggestion is very big. I have opened a new issue #250 to keep track of the suggestion. Let's wrap up this PR with the existing check() functionality.

prasadtalasila commented on util/environmentCheck.js:
```diff
@@ -0,0 +1,9 @@
+const check = function check(variable) {
```
Tests are missing for this function. You can also think of putting in the isValidNode() and InValidNodeException in this module itself.

prasadtalasila commented:
@AnkshitJain I am done with my review. I don't have any further suggestions to make. Everything else looks good for now.

AnkshitJain commented:
@prasadtalasila I have modified the method name and added the isValidNode method in status.js. I will add tests for environmentCheck.js along with the work concerning the custom error class. Please have a look at the updated commit.

prasadtalasila commented:
@AnkshitJain the latest commit looks good.

prasadtalasila commented:
@AnkshitJain please provide a tentative timeline to finish up with the pending work on this branch.

AnkshitJain commented:
@prasadtalasila Only the tests of environmentCheck.js are remaining, which I plan to add along with the custom error class. The tests will require adding some additional util functions. If the current PR is acceptable, it can be merged and closed, otherwise I will add the custom error class after 14th and it can be left open till then.

prasadtalasila commented:
@AnkshitJain any progress on this work?

prasadtalasila commented on load_balancer/load_balancer.js:
```diff
@@ -320,62 +288,29 @@ try {
-
 setInterval(function () {
   connection.query('SELECT 1' ,function(err, rows, fields) {
     console.log("keep alive query");
```
comment the console log line.

prasadtalasila commented on load_balancer/load_balancer.js:
```diff
@@ -320,62 +288,29 @@ try {
     console.log("keep alive query");
   });
-}, 10000);
+}, 100000);
```
The timeout value here would be half of the socket timeout set in MySQL config file.

AnkshitJain replied:
The default value for timeout is 8 hrs. We can have out own custom timeout value with the wait_timeout variable. It was verified and here is the same.
`````
mysql> SHOW global VARIABLES like "wait_timeout";
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| wait_timeout  | 28800 |
+---------------+-------+
1 row in set (0.07 sec)
`````      

We can set the timeout value to 14400 here.

prasadtalasila commented on tests/test.sh:
```diff
@@ -40,38 +40,46 @@ cp -f main_server/reval/reval.js tests/backup/main_server/
 # change the config file paths in all the relevant js files
-sed -i 's/\/etc\/load_balancer/\.\.\/deploy\/configs\/load_balancer/' load_balancer/load_balancer.js
-sed -i 's/\/etc\/main_server/\.\.\/deploy\/configs\/main_server/' main_server/main_server.js
+#sed -i 's/\/etc\/load_balancer/\.\.\/deploy\/configs\/load_balancer/' load_balancer/load_balancer.js
```
please remove the next two commented lines. There are equivalent environment config variables.

AnkshitJain commented:
@prasadtalasila The commit has been updated. Please have a look at it.
